### DISCUSSION TITLE
Zooming in plot X direction on startup or an event (like GUI button) - not manually

### DISCUSSION THREAD

**Discussion by Yazzito** [highfestiva/finplot#369] [General]
Hello, 

I am using finplot to create widgets to plot a stocks price (ochl). When it opens, the full data of the stock is shown on the widget/plot, as expected. I can scroll and zoom the widget manually and it all works. 

But what I would really like to have happen is this.:

when I create the plot that instead of seeing the full data, it automatically zooms in to the past years data, not all of the data. This is akin to me manually zooming in a few times and scrolling to the right so instead of looking at 30yrs worth of data, I zoom to the last years worth of data. 

Note: I'm not talking about removing items from the dataframe and replotting, as I want the data to still be there and I can manually scroll backwards as well. I just want to zoom in the Xdir automatically at startup instead of me having to manually zoom each plot.

Eventually, I would like to create GUI buttons that allow me to view the past 1yr, 3yr, 5yr, etc.. this is nearly identical to above but with different options and being triggered off a button press via a GUI event rather than at startup. Again, the GUI is easy to create but I don't know how to zoom?

in the finplot code, the zooming is tied in to mouse position, etc... but that isn't applicable in either of the cases above and I don't know how to trigger it externally.

```python
   def wheelEvent(self, ev, axis=None):
        if self.master_viewbox:
            return self.master_viewbox.wheelEvent(ev, axis=axis)
        if ev.modifiers() == QtCore.Qt.KeyboardModifier.ControlModifier:
            scale_fact = 1
            self.v_zoom_scale /= 1.02 ** (ev.delta() * self.state['wheelScaleFactor'])
        else:
            scale_fact = 1.02 ** (ev.delta() * self.state['wheelScaleFactor'])
        vr = self.targetRect()
        center = self.mapToView(ev.scenePos())
        pct_x = (center.x()-vr.left()) / vr.width()
        if pct_x < 0.05:  # zoom to far left => all the way left
            center = pg.Point(vr.left(), center.y())
        elif pct_x > 0.95:  # zoom to far right => all the way right
            center = pg.Point(vr.right(), center.y())
        self.zoom_rect(vr, scale_fact, center)
        # update crosshair
        _mouse_moved(self.win, None)
        ev.accept()
```

It would appear that these 4 variables inside FinViewBox are the key, but I'm not sure how to get at or use this data from inside my code. This seems to get called internally during a wheelevent, which I don't have so I was going to try and mimic it. 

1) vr = self.targetRect()  <- vr is the visible rectangle (QRectF) - which, I believe, in default is the full screen coordinates.
2) center = self.mapToView(ev.scenePos()) <- center (QPointF) which is the location of the mouse event press location. This isn't useful in my case because I don't have a mouse event. 
3) pct_x = (center.x()-vr.left()) / vr.width() <- pct_x is the % that is visible?
4) scale_fact = 1.02 ** (ev.delta() * self.state['wheelScaleFactor']) <- scale factor %

So my questions are:
1) Although I've looked through the code, perhaps I missed it. Is there a function to zoom in the X direction (time axis)?

2) If not, I was thinking of adding a new function to finplot to mimic what the wheelevent zoom code is doing but I'm not sure how to get or set the scale variables above. Any ideas on the best way to approach this? I'm not sure how to access the correct plot axs & FinViewBox parameters above. In the mouse wheel zoom, it's triggered off an event (ev) which appears to handle that.

PS. I did search the discussion forum here and there were a few topics regarding zoom but mostly on Y zooming, mouse wheel zoom or replotting after receiving new data. If there is a discussion on X zooming that I missed, sorry, please point me to it.

Thank you!!

**Comment by highfestiva** [comment]
Look at `examples/animate.py`.

**Comment by Yazzito** [comment]
Thank you! For future record, I was able to use the file that highfestiva mentioned (animate.py) to identify the line that does the horiz scrolling.

set_x_pos(xmin, xmax, ax=chart_item_here)
where xmin and xmax are the positions of the plot dataframe that you want to zoom in to.

The only gotcha is the granularity requires a certain amount of data points or it won't zoom at all.

Thank you again highfestiva!
